spec/features/admin/poll/shifts_spec.rb

```diff
@@ -48,6 +48,8 @@
       click_link "Manage shifts"
     end
 
+    visit new_admin_booth_shift_path booth_id: booth.id
```

Comment:
> Here was some strange behaviour -- click_link "Manage shifts" didn't follow the link, so I added `visit new_admin_booth_shift_path booth_id: booth.id`

DenisNikolski Oh, I see :thinking:. This is a very common issue; the first time you face it it's really hard to debug, but luckily from the second time it's very easy to fix.

The problem here is the `search` field is already present before clicking on "Manage shifts", and clicking this link makes an AJAX request. So the next line `fill_in "search", with: officer.email` might be executed before the AJAX request is finished. After that, anything can happen, since we're trying to do two things at the same time.

The solution is usually to insert an expectation to make sure the AJAX request is finished before filling in the field, which is the user behaviour we're trying to test.

Would you mind amending the last commit with this patch? :pray: This way we test the expected user behaviour of clicking links instead of entering URLs manually. Thanks a lot!

```diff
diff --git a/spec/features/admin/poll/shifts_spec.rb b/spec/features/admin/poll/shifts_spec.rb
index 484fdf5..34a7b81 100644
--- a/spec/features/admin/poll/shifts_spec.rb
+++ b/spec/features/admin/poll/shifts_spec.rb
@@ -48,7 +48,7 @@ describe "Admin shifts" do
       click_link "Manage shifts"
     end
 
-    visit new_admin_booth_shift_path booth_id: booth.id
+    expect(page).to have_content "This booth has no shifts"
 
     fill_in "search", with: officer.email
     click_button "Search"
@@ -74,7 +74,7 @@ describe "Admin shifts" do
       click_link "Manage shifts"
     end
 
-    visit new_admin_booth_shift_path booth_id: booth.id
+    expect(page).to have_css(".shift", count: 1)
 
     fill_in "search", with: officer.email
     click_button "Search"
@@ -119,7 +119,7 @@ describe "Admin shifts" do
       click_link "Manage shifts"
     end
 
-    visit new_admin_booth_shift_path booth_id: booth.id
+    expect(page).to have_css(".shift", count: 2)
 
     fill_in "search", with: officer.email
     click_button "Search"
@@ -160,7 +160,7 @@ describe "Admin shifts" do
       click_link "Manage shifts"
     end
 
-    visit new_admin_booth_shift_path booth_id: booth.id
+    expect(page).to have_content "This booth has no shifts"
 
     fill_in "search", with: officer.email
     click_button "Search"
```

Reply:
Here's the right patch:

```diff
diff --git a/spec/features/admin/poll/shifts_spec.rb b/spec/features/admin/poll/shifts_spec.rb
index 484fdf5..bbde7be 100644
--- a/spec/features/admin/poll/shifts_spec.rb
+++ b/spec/features/admin/poll/shifts_spec.rb
@@ -48,7 +48,7 @@ describe "Admin shifts" do
       click_link "Manage shifts"
     end
 
-    visit new_admin_booth_shift_path booth_id: booth.id
+    expect(page).to have_content "This booth has no shifts"
 
     fill_in "search", with: officer.email
     click_button "Search"
@@ -74,7 +74,7 @@ describe "Admin shifts" do
       click_link "Manage shifts"
     end
 
-    visit new_admin_booth_shift_path booth_id: booth.id
+    expect(page).to have_css(".shift", count: 1)
 
     fill_in "search", with: officer.email
     click_button "Search"
@@ -119,7 +119,7 @@ describe "Admin shifts" do
       click_link "Manage shifts"
     end
 
-    visit new_admin_booth_shift_path booth_id: booth.id
+    expect(page).to have_css(".shift", count: 2)
 
     fill_in "search", with: officer.email
     click_button "Search"
@@ -160,7 +160,7 @@ describe "Admin shifts" do
       click_link "Manage shifts"
     end
 
-    visit new_admin_booth_shift_path booth_id: booth.id
+    expect(page).to have_content "This booth has no shifts"
 
     fill_in "search", with: officer.email
     click_button "Search"
@@ -184,8 +184,6 @@ describe "Admin shifts" do
       click_link "Manage shifts"
     end
 
-    visit new_admin_booth_shift_path booth_id: booth.id
-
     expect(page).to have_css(".shift", count: 1)
     within("#shift_#{shift.id}") do
       click_link "Remove"
@@ -210,8 +208,6 @@ describe "Admin shifts" do
       click_link "Manage shifts"
     end
 
-    visit new_admin_booth_shift_path booth_id: booth.id
-
     expect(page).to have_css(".shift", count: 1)
     within("#shift_#{shift.id}") do
       click_link "Remove"
@@ -239,8 +235,6 @@ describe "Admin shifts" do
       click_link "Manage shifts"
     end
 
-    visit new_admin_booth_shift_path booth_id: booth.id
-
     expect(page).to have_css(".shift", count: 1)
     within("#shift_#{shift.id}") do
       click_link "Remove"
```

Reply:
Cool, thanks for explanation